### ISSUE TITLE
rust::Str and rust::String not being null terminated is not obvious

### ISSUE THREAD

**Issue by TheZoq2** [dtolnay/cxx#224]
First of all, thanks for this great library, with very little effort, it allows me to write a nice rust frontend fro my c++ project.

When I started using it to pass some strings from the rust side, I quickly found the Builtin types table, and given that a `data` function is available on `rust::Str`, I assumed that it would be fine to copy the content using `std::string(rust_string.data())`. Obviously this turned out to not be the case, but I don't think that's clear, at least from the parts of the documentation that I have read.

Would it be a good idea to put an entry in the "restrictions" column saying that this is the case?

**Comment by dtolnay** [comment]
The intended reference for the API of the C++ types is the header: https://github.com/dtolnay/cxx/blob/0.3.4/include/cxx.h (for now; at least until #179 is closed). This is probably where you found the `data` function.

The header has a comment on Str::data and String::data indicating that there is no null terminator:

https://github.com/dtolnay/cxx/blob/bbd2620921e921825575810a9fd40712c378d6aa/include/cxx.h#L37-L38

https://github.com/dtolnay/cxx/blob/bbd2620921e921825575810a9fd40712c378d6aa/include/cxx.h#L66-L67

**Comment by TheZoq2** [comment]
> This is probably where you found the data function.
Not quite, I did find it when noticing a problem with the strings, but I initially found it using my editor autocomplete. I guess this issue is mostly a "I read the bare minimum amount of docs and got surprised when things went wrong :)"

**Comment by xfix** [comment]
`rust::Str` is like `std::string_view` which is not guaranteed to be null terminated.

**Comment by cgwalters** [comment]
I'm actually converting a C codebase to "compile in C++ mode" purely so I can use this project.  And starting from a world wher e NUL terminated C strings are pervasive then adding C++ `std::string` and Rust `String` is very painful so far (basically need to do `auto s = std::string(rustobject->getstring()); some_c_function(s.c_str())`.

In some cases on the Rust side we've been using https://crates.io/crates/c_utf8 (which...IMO should probably be part of Rust std but that's a whole other discussion).  Using that one can avoid a copy.  Perhaps this crate could have a feature gate to enable adding that as a binding?

**Comment by dtolnay** [comment]
How about adding a c_str() member function for rust::String? #586
